lib/oli/publishing.ex

```diff
@@ -869,6 +864,42 @@ defmodule Oli.Publishing do
     end)
   end
 
+  # For a given publication, gather all of the part ids and their grading approach
+  # from within the revisions of activities.  Inserts the three element tuple
+  # of {part_id, grading_approach, revision_id} into the `revision_parts` table.
+  #
+  # There will be conflicts, of course, as later publications will have the same
+  # three element tuples.  The ON CONFLICT ... DO NOTHING handles this.
+  #
+  # This replaces the materialized view approach of the "part_mapping", but in a way
+  # that is far more efficient since it operates against a single publication,
+  # where the part_mapping refresh operated over the entire published_resources table.
+  #
+  def insert_revision_part_records(publication_id) do
+    query = """
+      INSERT INTO revision_parts(part_id, grading_approach, revision_id)
+      SELECT DISTINCT t.parts->>'id' as part_id, t.parts->>'gradingApproach' as grading_approach, t.revision_id as revision_id FROM (
+        SELECT jsonb_path_query(r.content, '$."authoring"."parts"[*]') as parts,
+          r.id as revision_id
+        FROM published_resources pr
+          LEFT JOIN publications p ON p.id = pr.publication_id
+          LEFT JOIN revisions r ON r.id = pr.revision_id
+        WHERE pr.publication_id = $1 AND r.resource_type_id = 3) t
```

Comment:
Maybe we could get the `resource_type_id` dynamically

Reply:
Ah yes indeed we could.  Specifying a param here and supplying it with `Oli.Resources.ResourceType.get_id_by_type("activity")` would do the trick

Reply:
I filed https://eliterate.atlassian.net/browse/MER-1609 for tracking this work